pkg/event/registry_test.go

```diff
@@ -0,0 +1,106 @@
+package event_test
+
+import (
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	k8sV1 "k8s.io/api/core/v1"
+	v1meta "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	submV1 "github.com/submariner-io/submariner/pkg/apis/submariner.io/v1"
+	"github.com/submariner-io/submariner/pkg/event"
+	"github.com/submariner-io/submariner/pkg/event/logger"
+)
+
+const npOVNKubernetes = "OVNKubernetes"
+const npGenericKubeproxyIptables = "GenericKubeproxyIptables"
+
+var _ = Describe("Event Registry", func() {
+	When("Registry is populated with different network plugin handlers", func() {
+		var (
+			ovn, generic *TestHandler
+			registry     event.Registry
+		)
+
+		BeforeEach(func() {
+			registry = event.NewRegistry("test-registry", npGenericKubeproxyIptables)
+			ovn = NewTestHandler("handler-ovn", npOVNKubernetes)
+			generic = NewTestHandler("handler-ovn", npGenericKubeproxyIptables)
+			registry.AddHandlers(logger.NewHandler(), ovn, generic)
+		})
+
+		It("Should only init handlers of the current network plugin", func() {
+			Expect(ovn.Initialized).To(BeFalse())
+			Expect(generic.Initialized).To(BeTrue())
+		})
+
+		It("Should only call handlers of the current network plugin", func() {
+			_ = registry.TransitionToGateway()
+			Expect(ovn.EventList).To(HaveLen(0))
+			Expect(generic.EventList).To(HaveLen(1))
+		})
+	})
+
+	When("Registry is populated with same plugin handlers", func() {
+		var (
+			handler1, handler2 *TestHandler
+			registry           event.Registry
+		)
+
+		BeforeEach(func() {
+			registry = event.NewRegistry("test-registry", npGenericKubeproxyIptables)
+			handler1 = NewTestHandler("handler1", npGenericKubeproxyIptables)
+			handler2 = NewTestHandler("handler2", npGenericKubeproxyIptables)
+			registry.AddHandlers(logger.NewHandler(), handler1, handler2)
+		})
+
+		It("All handlers should be initialized", func() {
+			Expect(handler1.Initialized).To(BeTrue())
+			Expect(handler2.Initialized).To(BeTrue())
+		})
+
+		It("Should call all handlers", func() {
+			_ = registry.TransitionToGateway()
+			Expect(handler1.EventList).To(HaveLen(1))
+			Expect(handler2.EventList).To(HaveLen(1))
+			Expect(handler1.EventList[0].Name).Should(Equal(evTransitionToGateway))
+			Expect(handler2.EventList[0].Name).Should(Equal(evTransitionToGateway))
+		})
+
+		It("Should be called in registration order", func() {
+			_ = registry.TransitionToGateway()
+			Expect(handler1.EventList).To(HaveLen(1))
+			Expect(handler2.EventList).To(HaveLen(1))
+			Expect(handler1.EventList[0].Timestamp).To(BeNumerically("<", handler2.EventList[0].Timestamp))
```

Comment:
Well, if we have processors running under 1ns per instruction, and al that can happen on one instruction.
I can change the Timestamp to be a global incremented variable, instead of a global chan. The chan is elegant TBH, but I'd prefer not to spend too much time in making the test more elgant, this is blocking a lot of work ':)